staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/convertlastapplied.go

```diff
@@ -0,0 +1,150 @@
+}
+
+// Apply implements Manager.
+func (f *convertLastAppliedManager) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	//TODO: fail if there's a conversion error?
+	liveObj, managed, _ = f.convertLastAppliedToFieldManager(liveObj, managed, manager)
+
+	newObj, managed, err := f.fieldManager.Apply(liveObj, newObj, managed, manager, force)
+	if err != nil {
+		return nil, nil, err
+	}
+	return newObj, managed, nil
+}
+
+func (f *convertLastAppliedManager) convertLastAppliedToFieldManager(liveObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	var accessor, err = meta.Accessor(liveObj)
+	if err != nil {
+		panic(fmt.Sprintf("couldn't get accessor: %v", err))
+	}
+
+	var annotations = accessor.GetAnnotations()
+	if annotations == nil {
+		return liveObj, nil, nil
+	}
+
+	var lastApplied, ok = annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok || lastApplied == "" {
+		return liveObj, nil, nil
+	}
+
+	liveObjTyped, err := f.typeConverter.ObjectToTyped(liveObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert live obj to typed: %v", err)
+	}
+
+	liveObjFieldSet, err := liveObjTyped.ToFieldSet()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create fieldset for live obj: %v", err)
+	}
+
+	var lastAppliedObj = &unstructured.Unstructured{Object: map[string]interface{}{}}
+	err = json.Unmarshal([]byte(lastApplied), lastAppliedObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to decode last applied obj: %v", err)
+	}
+
+	//TODO: why would this ever happen?
+	internal.RemoveObjectManagedFields(lastAppliedObj)
+
+	//TODO: is this version check necessary?
+	if lastAppliedObj.GetAPIVersion() != f.groupVersion.String() {
+		return nil, nil, fmt.Errorf("expected version of last applied to match live object '%s', but got '%s': %v", f.groupVersion.String(), lastAppliedObj.GetAPIVersion(), err)
+	}
+
+	lastAppliedObjTyped, err := f.typeConverter.ObjectToTyped(lastAppliedObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert last applied to typed: %v", err)
+	}
+
+	lastAppliedObjFieldSet, err := lastAppliedObjTyped.ToFieldSet()
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create fieldset for last applied object: %v", err)
+	}
+
+	comparison, err := lastAppliedObjTyped.Compare(liveObjTyped)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to compare last applied object and live object: %v", err)
+	}
+
+	lastAppliedObjFieldSet = lastAppliedObjFieldSet.
+		Difference(comparison.Modified).
+		Difference(comparison.Removed)
+
+	var apiVersion = fieldpath.APIVersion(f.groupVersion.String())
+
+	var intersection = lastAppliedObjFieldSet.Intersection(liveObjFieldSet)
+	if !intersection.Empty() {
+		managed.Fields()[manager] = fieldpath.NewVersionedSet(intersection, apiVersion, false)
+	}
+
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
```

Comment:
I don't think we want to delete this anymore since we currently want to keep this up to date